Include image matches in find-in-page results on iOS Safari

danae404 opened the pull request:
#### 7bdaa877929166c898c7d2d221249ea2d0ef0ef9
<pre>
Include image matches in find-in-page results on iOS Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=243948">https://bugs.webkit.org/show_bug.cgi?id=243948</a>
rdar://problem/98681935

Reviewed by NOBODY (OOPS!).

No tests yet.

This patch improves the internal setting ImageAnalysisForFindInPage, which
Currently only supports macOS Safari. With this patch, iOS Safari find-in-page
follows the same behavior as macOS Safari, where all images on the page are
enqueued, and when the queue empties, the most recent find-in-page is repeated
to include any matches from the most recently analyzed images.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateFindResults): Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFindResults): Added.
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.h:

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:

(WebKit::WebFoundTextRangeController::updateFindResults): Added.

Initiates IPC that eventually calls [findInteraction presentFindNavigatorShowingReplace:NO],
Which triggers the most recent search to repeat.

(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):

Calls Page::analyzeImagesForFindInPage with WebFoundTextRangeController::updateFindResults as a callback to be called when the image queue empties.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateFindResults): Added.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
</pre>

pxlcoder commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -465,6 +465,7 @@ class PageClient : public CanMakeWeakPtr<PageClient> {
     virtual void restorePageCenterAndScale(std::optional<WebCore::FloatPoint> center, double scale) = 0;
 
     virtual void elementDidFocus(const FocusedElementInformation&, bool userIsInteracting, bool blurPreviousNode, OptionSet<WebCore::ActivityState::Flag> activityStateChanges, API::Object* userData) = 0;
+    virtual void presentFindNavigatorWithoutReplace() = 0;
```
I think we should rename this method (and the ones below) to something like `updateFindResults`, since we're really working around the fact that UIKit doesn't currently expose a better way to update results, and the find navigator is already visible.

pxlcoder replied:
I would move this declaration further down, since the rest of the methods here are about element focus.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -52,8 +52,18 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
 void WebFoundTextRangeController::findTextRangesForStringMatches(const String& string, OptionSet<FindOptions> options, uint32_t maxMatchCount, CompletionHandler<void(Vector<WebFoundTextRange>&&)>&& completionHandler)
 {
+#if ENABLE(IMAGE_ANALYSIS)
+    m_webPage->corePage()->analyzeImagesForFindInPage([this] {
```
This should capture a `WeakPtr`, there's no guarantee the `WebFoundTextRangeController` will still be alive.

Will also need to make the class inherit from `public CanMakeWeakPtr< WebFoundTextRangeController>`.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -52,8 +52,21 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
 void WebFoundTextRangeController::findTextRangesForStringMatches(const String& string, OptionSet<FindOptions> options, uint32_t maxMatchCount, CompletionHandler<void(Vector<WebFoundTextRange>&&)>&& completionHandler)
 {
+#if ENABLE(IMAGE_ANALYSIS) && PLATFORM(IOS_FAMILY)
+    m_webPage->corePage()->analyzeImagesForFindInPage([weakController = WeakPtr { this }] {
```
Nit: We normally name the variable `weakThis`.

cdumez replied:
also `weakThis = WeakPtr { *this }` (note the `*this`)

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -52,8 +52,21 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
 {
 }
 
+#if PLATFORM(IOS_FAMILY)
```
I've tried to avoid making this class have `IOS_FAMILY` specific logic. Can you try to remove `IOS_FAMILY` where possible? The shared logic could have an empty implementation.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -52,8 +52,21 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
+        if (weakController)
+            weakController->updateFindResults();
+    });
+#endif
```
Nit: Newline after this `#endif`.

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -636,6 +636,12 @@ - (BOOL)_isHostedInAnotherProcess;
 
+void PageClientImpl::updateFindResults()
+{
+    auto findInteraction = [m_webView findInteraction];
+    [findInteraction presentFindNavigatorShowingReplace:NO];
```
I think this needs to be guarded by `HAVE(UIFINDINTERACTION)`.

pxlcoder commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -337,6 +337,8 @@ class PageClient : public CanMakeWeakPtr<PageClient> {
     virtual WebCore::IntPoint accessibilityScreenToRootView(const WebCore::IntPoint&) = 0;
     virtual WebCore::IntRect rootViewToAccessibilityScreen(const WebCore::IntRect&) = 0;
+
+    virtual void updateFindResults() = 0;
```
This can be `virtual void updateFindResults() { }`, and then you wouldn't need to implement it everywhere.

danae404 committed:
Include image matches in find-in-page results on iOS Safari
https://bugs.webkit.org/show_bug.cgi?id=243948
rdar://problem/98681935

Reviewed by NOBODY (OOPS!).

No tests yet.

This patch improves the internal setting ImageAnalysisForFindInPage, which
Currently only supports macOS Safari. With this patch, iOS Safari find-in-page
follows the same behavior as macOS Safari, where all images on the page are
enqueued, and when the queue empties, the most recent find-in-page is repeated
to include any matches from the most recently analyzed images.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::updateFindResults): Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateFindResults): Added.
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.h:

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:

(WebKit::WebFoundTextRangeController::updateFindResults): Added.

Initiates IPC that eventually calls [findInteraction presentFindNavigatorShowingReplace:NO],
Which triggers the most recent search to repeat.

(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):

Calls Page::analyzeImagesForFindInPage with WebFoundTextRangeController::updateFindResults as a callback to be called when the image queue empties.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.h:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateFindResults): Added.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

whsieh reviewed (commented):
Seems quite reasonable!

A simple test would be nice addition — there are some examples in `FindInPage.mm`, which you could use as reference.

darinadler reviewed (changes requested):
Please update to include a test for this as @whsieh suggested and then we can approve and merge this. Thank you for contributing this!
